include/ginkgo/core/base/array.hpp

```diff
@@ -547,6 +547,61 @@ class array {
         return *this;
     }
 
+    /**
+     * Copies or converts data from a const_array_view.
+     *
+     * In the case of an array target, the array is resized to match the
+     * source's size. In the case of a view target, if the dimensions are not
+     * compatible a gko::OutOfBoundsError is thrown.
+     *
+     * This does not invoke the constructors of the elements, instead they are
+     * copied as POD types.
+     *
+     * The executor of this is preserved. In case this does not have an assigned
+     * executor, it will inherit the executor of other.
+     *
+     * @param other  the const_array_view to copy from
+     * @tparam OtherValueType  the value type of `other`
+     *
+     * @return this
+     */
+    template <typename OtherValueType>
+    std::enable_if_t<std::is_convertible<OtherValueType, ValueType>::value,
+                     array>&
+    operator=(const detail::const_array_view<OtherValueType>& other)
```

Comment:
not sure I like this - I think we should be using `resize_and_reset` and `exec->copy_from` explicitly, because it mixes the behavior between assigning a view and a const view.

Reply:
I don't think it's mixing behavior. In case of a non-const view, `a = view_b` also does a copy. Only the `a = std::move(view_b)` is different.

Reply:
That's what you are doing in your `read(const device_matrix_data)` function - we have rvalue overloads for regular arrays, but not for const views.

Reply:
I'm not sure what you are getting at. Should I also add move assignment for const views?

Reply:
I mean if we cannot match the semantics we have for array views with const_array_view (which is impossible without const_cast, and thus unsafe), I think we should avoid expanding the use of `const_array_view` beyond it being a wrapper type we can use in conjunction with `create_const`

Reply:
IMO copying a `const_array_view` into an array is a perfectly fine use case. Otherwise copying from a `const T*` into an array is quite complicated and involves a lot of boilerplate. Basically each time you want to do such a copy, you have to copy-paste the implemented copy-assigment to catch all edge cases.

I would also go further and allow for copy-construction.

Reply:
I think this will be fine, the distinction between move-assignment and copy-assignment is probably not that important for any but very advanced users, and they should be able to figure out the difference between a const and non-const rhs.